File: src/main/java/io/scriptor/result/StreamResult.java
package io.scriptor.result;

import java.io.InputStream;

public class StreamResult extends ResultBase<InputStream> {

    public StreamResult(final int statusCode) {
        super(statusCode);
    }

    public StreamResult(final int statusCode, final String statusText) {
        super(statusCode, statusText);
    }

    public StreamResult(final int statusCode, final String statusText, final InputStream value) {
        super(statusCode, statusText, value);
    }

    @Override
    public int getSize() {
        return -1;
    }

    @Override
    public InputStream getStream() {
        return getValue();
    }
}

File: server/src/main/kotlin/dev/scriptor/server/result/StreamResult.kt
package dev.scriptor.server.result

import dev.scriptor.server.ParameterList
import dev.scriptor.server.RangeReadableByteChannel
import java.io.InputStream
import java.nio.channels.Channels

class StreamResult : Result {

    constructor(
        statusCode: Int = 200,
        statusText: String = "OK",
        contentType: String = "application/octet-stream",
        headers: ParameterList = ParameterList(),
        value: InputStream,
        count: Long = -1L,
    ) : super(
        statusCode,
        statusText,
        contentType,
        headers,
        when {
            count < 0L -> Channels.newChannel(value)
            else -> RangeReadableByteChannel(Channels.newChannel(value), 0L until count)
        },
    )
}
